duktape/src/main/jni/DuktapeContext.cpp

```diff
@@ -0,0 +1,160 @@
+  duk_get_prop_string(ctx, -1, JAVA_METHOD_PROP_NAME);
+  JavaMethod* method = static_cast<JavaMethod*>(duk_require_pointer(ctx, -1));
+  duk_pop_2(ctx);
+  return method;
+}
+
+duk_int_t eval_string_with_filename(duk_context *ctx, const char *src,
+                                    const char *fileName) {
+  duk_push_string(ctx, fileName);
+  return duk_eval_raw(ctx, src, 0, DUK_COMPILE_EVAL | DUK_COMPILE_SAFE |
+                                   DUK_COMPILE_NOSOURCE | DUK_COMPILE_STRLEN);
+}
+
+duk_ret_t handler(duk_context *ctx) {
+  JavaMethod* method = getJavaMethod(ctx);
+  return method ? method->Invoke(ctx, getJNIEnv(ctx), getJavaThis(ctx)) : DUK_RET_INTERNAL_ERROR;
+}
+
+duk_ret_t javaObjectFinalizer(duk_context *ctx) {
+  jobject thisObject = static_cast<jobject>(duk_require_pointer(ctx, -1));
+  getJNIEnv(ctx)->DeleteGlobalRef(thisObject);
+  duk_enum(ctx, -1, DUK_ENUM_OWN_PROPERTIES_ONLY);
+  while (duk_next(ctx, -1, true)) {
+    if (duk_get_prop_string(ctx, -1, JAVA_METHOD_PROP_NAME)) {
+      delete static_cast<JavaMethod*>(duk_require_pointer(ctx, -1));
+    }
+    duk_pop_3(ctx);
+  }
+  duk_pop_2(ctx);
+  return 0;
+}
+
+} // anonymous namespace
+
+Duktape::Duktape(JavaVM* javaVM)
+    : m_context(duk_create_heap_default()) {
+  // Stash the JVM object in the context, so we can find our way back from a duktape C callback.
+  duk_push_global_stash(m_context);
+  duk_push_pointer(m_context, javaVM);
+  duk_put_prop_string(m_context, -2, JAVAVM_PROP_NAME);
+  duk_pop(m_context);
+}
+
+Duktape::~Duktape() {
+  duk_destroy_heap(m_context);
+}
+
+jstring Duktape::evaluate(JNIEnv *env, jstring code, jstring fname) {
+  const JString sourceCode(env, code);
+  const JString fileName(env, fname);
+
+  jstring result = nullptr;
+
+  if (eval_string_with_filename(m_context, sourceCode, fileName) != 0) {
+    jclass exceptionClass = env->FindClass("com/squareup/duktape/DuktapeException");
+
+    // If it's a duktape error object, try to pull out the full stacktrace.
+    if (duk_is_error(m_context, -1) && duk_get_prop_string(m_context, -1, "stack")) {
+      const char* stack = duk_safe_to_string(m_context, -1);
+
+      // Is there an exception thrown from a Java method?
+      if (duk_get_prop_string(m_context, -2, JavaMethod::JAVA_EXCEPTION_PROP_NAME)) {
+        // TODO: add the duktape stack to this exception.
+        jthrowable ex = static_cast<jthrowable>(duk_get_pointer(m_context, -1));
+        env->Throw(ex);
+        duk_pop(m_context);
+      } else {
+        env->ThrowNew(exceptionClass, stack);
+      }
+      duk_pop(m_context);
+    } else {
+      // Not an error or no stacktrace, just convert to a string.
+      env->ThrowNew(exceptionClass, duk_safe_to_string(m_context, -1));
+    }
+  } else {
+    result = env->NewStringUTF(duk_get_string(m_context, -1));
+  }
+
+  duk_pop(m_context);
+
+  return result;
+}
+
+void Duktape::bindInstance(JNIEnv *env, jstring instance, jobject object, jobjectArray methods) {
```

Comment:
Yes, I'd still have to forward declare the javaObjectFinalizer above the use.  If it's all the same, I'd personally rather keep it as is.